packages/SwingSet/docs/vat-environment.md

```diff
@@ -0,0 +1,113 @@
+
+### Direct Eval Expressions
+
+A "direct eval", invoked like `eval(code)`, is defined to behave as if `code` were expanded in place. The evaluated code sees the same scope as the `eval` itself sees, so the `code` in:
+
+```js
+function foo(code) {
+  const x = 1;
+  eval(code);
+}
+```
+
+gets to reference `x`. If you perform a direct eval, you cannot hide your internal authorities from the code you're evaluating.
+
+In contrast, an "indirect eval" only gets the global scope, not the local scope. In a safe SES environment, indirect eval is a useful and common tool. The evaluated code can only access global objects, and those are all safe (and frozen). The only bad thing an indirect eval can do is consume unbounded CPU or memory. Once you've evaluated that code, you can give it as many or as few authories as you like, by invoking it with arguments.
+
+The most common way to invoke an indirect eval is like `(1,eval)(code)`.
+
+The SES shim cannot correctly emulate a "direct eval". If it tried, it would wind up performing an indirect eval. This could be pretty confusing, because the code might not actually use objects from the local scope, so you might not notice the problem until some future change altered the behavior.
+
+To avoid confusion, the shim use a regular expression to reject code that looks like it is performing a direct eval. This regexp is not complete (it is possible to trick it into performing a direct eval anyway), but that's ok, because our goal is merely to guide people away from confusing behaviors early in their development process.
+
+This regexp will falsely reject occurrences inside static strings and comments.
+
+### Frozen globalThis
+
+Vats run in a Compartment whose `globalThis` object is frozen. A mutable global object would provide an "ambient communication channel": one side sets `globalThis.heyBuddyAreYouOutThere = 'exfiltrated message'` and the other side periodically reads it. This would violate one of the rules of object-capability security: objects may only communicate through references.
+
+Vats can create new Compartments, and they get to decide whether the new `globalThis` is frozen or not. If the Compartment is meant to support ocap security, the creator should do `harden(c.globalThis)` before loading any code into it.
+
+### Frozen Primordials
+
+SES freezes the "primordials": the built-in JavaScript objects like `Object`, `Array`, and `RegExp`, as well as their prototype chains. This prevents malicious code from changing the behavior of built-ins in surprising ways (imagine `Array.prototype.push` being changed to deliver a copy of its argument to the attacker, or simply ignore certain values). It also prevents the use of e.g. `Object.heyBuddy` as an ambient communication channel.
```

Comment:
Italicize primordials.

"like" -> "such as"

"of e.g." -> "of, for example,"